doc/source/architecture.rst

```diff
@@ -0,0 +1,213 @@
+customization, it is necessary to Read the data model. Automation alone is not
+sufficient. For example, when integrating a third-party solver, you will need
+to access all of the boundary conditions, geometry, material properties, mesh,
+and connections to properly input them to the solver. As discussed above, the
+command pattern does not allow this kind of access pattern. Due to this fact,
+the Mechanical API exposes its data model directly to the user. This is how a
+hypothetical command-based API would look for renaming an object.
+
+.. code:: python
+
+    RenameCommand(Id=100, Name="New name")
+
+Instead, an API based on a data model, like mechanical, looks like this:
+
+.. code:: python
+
+    obj = DataModel.GetObjectById(100)
+    print(obj.Name)
+    obj.Name = "New name"
+
+Notice that the old name could be printed by accessing a property of the
+object. This has no equivalent if the API used a command pattern.
+
+
+PyMechanical remote interface
+-----------------------------
+
+One way to interact with the Mechanical API from CPython is as a Remote
+Session. A session of Mechanical runs as a server and can receive a Remote
+Procedure Call (RPC). The RPC is handled by the server and then the response
+is returned back to the client.
+
+PyMechanical embedded instance
+--------------------------------
+
+The embedded instance used by PyMechanical embeds an entire instance of the
+Mechanical application in-memory inside of a CPython program. There is no
+additional running process associated with it. Here, only the CPython interface
+to the Mechanical API can be used.
+
+Distributed systems
+===================
+
+This section contains a very basic explanation of Distributed Systems. It is
+not meant be exhaustive and rigorous, but instead introduces just the topics
+necessary to understand the choices made by PyMechanical.
+
+The term Distributed System is used to describe a software system that uses
+a network to distribute software across multiple physical machines. With a
+Distributed System, the individual pieces of that system do not share an
+address space, and therefore cannot call functions of each other directly.
+Instead, they communicate with each other by sending messages in packets.
+
+There are unique characteristics of distributed systems relative to classical
+software systems that share an address space. For instance, in a distributed
+system, any call can fail because of a problem with the network, and the caller
+can not always know whether a call has failed. For this reason, features of
+interest to distributed system designers such as fault tolerance and
+idempotency are not necessarily important in classical software systems.
+
+These systems are used by email, multiplayer games, web applications, and high
+performance computing, among other things. Historically, there are three major
+categories of remote APIs, Message passing, resource-based, and Remote Method
+Invocation.
+
```

Comment:
Examples of Distributed systems include email servers, web applications, and high-performance computing. Historically, there are three major categories of remote APIs, Message passing, resource-based, and Remote Method Invocation.